p2pclient/dht.go

```diff
@@ -0,0 +1,350 @@
+package p2pclient
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+
+	ggio "github.com/gogo/protobuf/io"
+	cid "github.com/ipfs/go-cid"
+	crypto "github.com/libp2p/go-libp2p-crypto"
+	pb "github.com/libp2p/go-libp2p-daemon/pb"
+	peer "github.com/libp2p/go-libp2p-peer"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// PeerInfo wraps the PeerInfo message from our protobuf with richer types.
+type PeerInfo struct {
+	// Id is the peer's ID
+	ID peer.ID
+	// Addrs are the peer's listen addresses.
+	Addrs []ma.Multiaddr
+}
+
+func convertPbPeerInfo(pbi *pb.PeerInfo) (*PeerInfo, error) {
```

Comment:
since we are passing PeerInfo by value now, let's do it here too and avoid heap allocation.
So return `PeerInfo` instead of pointer?